crates/ibc/src/services/core/client.rs

```diff
@@ -0,0 +1,314 @@
+use ibc_proto::ibc::core::client::v1::{
+    ConsensusStateWithHeight, IdentifiedClientState, Params as ClientParams,
+    QueryClientParamsRequest, QueryClientParamsResponse, QueryClientStateRequest,
+    QueryClientStateResponse, QueryClientStatesRequest, QueryClientStatesResponse,
+    QueryClientStatusRequest, QueryClientStatusResponse, QueryConsensusStateHeightsRequest,
+    QueryConsensusStateHeightsResponse, QueryConsensusStateRequest, QueryConsensusStateResponse,
+    QueryConsensusStatesRequest, QueryConsensusStatesResponse, QueryUpgradedClientStateRequest,
+    QueryUpgradedClientStateResponse, QueryUpgradedConsensusStateRequest,
+    QueryUpgradedConsensusStateResponse,
+};
+use tonic::{Request, Response, Status};
+use tracing::trace;
+
+use crate::core::ics02_client::error::ClientError;
+use crate::core::ics24_host::identifier::ClientId;
+use crate::core::ics24_host::path::{
+    ClientConsensusStatePath, ClientStatePath, Path, UpgradeClientPath,
+};
+use crate::core::{ContextError, ValidationContext};
+use crate::hosts::tendermint::upgrade_proposal::UpgradeValidationContext;
+use crate::prelude::*;
+use crate::services::core::context::QueryContext;
+use crate::Height;
+
+pub struct ClientQueryService<I, U>
+where
+    I: QueryContext + Send + Sync + 'static,
+    U: UpgradeValidationContext + Send + Sync + 'static,
+    <I as ValidationContext>::AnyClientState: Into<Any>,
+    <I as ValidationContext>::AnyConsensusState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyClientState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyConsensusState: Into<Any>,
+{
+    ibc_context: I,
+    upgrade_context: U,
+}
+
+impl<I, U> ClientQueryService<I, U>
+where
+    I: QueryContext + Send + Sync + 'static,
+    U: UpgradeValidationContext + Send + Sync + 'static,
+    <I as ValidationContext>::AnyClientState: Into<Any>,
+    <I as ValidationContext>::AnyConsensusState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyClientState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyConsensusState: Into<Any>,
+{
+    pub fn new(ibc_context: I, upgrade_context: U) -> Self {
+        Self {
+            ibc_context,
+            upgrade_context,
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl<I, U> ClientQuery for ClientQueryService<I, U>
+where
+    I: QueryContext + Send + Sync + 'static,
+    U: UpgradeValidationContext + Send + Sync + 'static,
+    <I as ValidationContext>::AnyClientState: Into<Any>,
+    <I as ValidationContext>::AnyConsensusState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyClientState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyConsensusState: Into<Any>,
+{
+    async fn client_state(
+        &self,
+        request: Request<QueryClientStateRequest>,
+    ) -> Result<Response<QueryClientStateResponse>, Status> {
+        let request_ref = request.get_ref();
+
+        trace!("Got client state request: {:?}", request_ref);
```

Comment:
Can we drop these traces and let the implementors handle tracing themselves? They can include their tracing system right before or after their gRPC calls. Hosts may not necessarily use `tracing` dependency, which means they can't capture this log in that manner.

Reply:
Users don't have access to these methods. These are called inside the tonic grpc server handler. But I understand your point about `tracing` dependency. Not sure what to do here.

Reply:
It's actually a good idea to include traces; it is standard in networking libraries to do so. For some reason we already had `tracing` as a dependency. Not sure if it's used anywhere else, but we could have it as a dependency only for the `grpc` feature

Reply:
@plafer, I discussed this with @Farhad-Shabani, and I agree with him now. I didn't know that it was not being used by ibc-rs thoroughly. So, using it here suddenly breaks the code repository norm. Let's merge the PR without them. If a client requests tracing, we can discuss it in future issues.